File: services/api/src/instrumentation.ts
/**
 * OpenTelemetry instrumentation setup
 * Configures tracing, metrics, and log correlation for the API service
 */
import { NodeSDK } from "@opentelemetry/sdk-node";
import { getNodeAutoInstrumentations } from "@opentelemetry/auto-instrumentations-node";
import { OTLPTraceExporter } from "@opentelemetry/exporter-trace-otlp-http";
import { OTLPMetricExporter } from "@opentelemetry/exporter-metrics-otlp-http";
import { Resource } from "@opentelemetry/resources";
import {
  SEMRESATTRS_SERVICE_NAME,
  SEMRESATTRS_SERVICE_VERSION,
} from "@opentelemetry/semantic-conventions";
import { PeriodicExportingMetricReader } from "@opentelemetry/sdk-metrics";
import { BatchSpanProcessor } from "@opentelemetry/sdk-trace-base";

/**
 * Initialize OpenTelemetry SDK
 * Must be called before any other imports that use OpenTelemetry
 */
export function initializeTelemetry(): void {
  const serviceName = process.env.OTEL_SERVICE_NAME || "api";
  const serviceVersion = process.env.API_VERSION || "1.0.0";
  const otlpEndpoint =
    process.env.OTEL_EXPORTER_OTLP_ENDPOINT || "http://otel-collector:4317";
  const environment = process.env.DD_ENV || "demo";
  const customerTier = process.env.CUSTOMER_TIER || "premium";
  const region = process.env.REGION || "us-east-1";

  // Resource attributes (tags that apply to all telemetry)
  const resource = new Resource({
    [SEMRESATTRS_SERVICE_NAME]: serviceName,
    [SEMRESATTRS_SERVICE_VERSION]: serviceVersion,
    "deployment.environment": environment,
    customer_tier: customerTier,
    region: region,
  });

  // OTLP exporters
  const traceExporter = new OTLPTraceExporter({
    url: `${otlpEndpoint}/v1/traces`,
  });

  const metricExporter = new OTLPMetricExporter({
    url: `${otlpEndpoint}/v1/metrics`,
  });

  // Initialize SDK
  const sdk = new NodeSDK({
    resource,
    spanProcessor: new BatchSpanProcessor(traceExporter),
    metricReader: new PeriodicExportingMetricReader({
      exporter: metricExporter,
      exportIntervalMillis: 10000, // Export metrics every 10 seconds
    }),
    instrumentations: [
      getNodeAutoInstrumentations({
        // Enable auto-instrumentation for common libraries
        "@opentelemetry/instrumentation-fs": {
          enabled: false, // Disable file system instrumentation for cleaner traces
        },
      }),
    ],
  });

  sdk.start();

  console.log(
    `OpenTelemetry initialized for ${serviceName} v${serviceVersion}`,
  );

  // Graceful shutdown
  process.on("SIGTERM", () => {
    sdk
      .shutdown()
      .then(() => console.log("OpenTelemetry terminated"))
      .catch((error) => console.error("Error terminating OpenTelemetry", error))
      .finally(() => process.exit(0));
  });
}

File: services/worker/app/instrumentation.py
"""
OpenTelemetry instrumentation setup for Python worker
Configures tracing and metrics for Celery tasks
"""
import os
from opentelemetry import trace, metrics
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor
from opentelemetry.sdk.metrics import MeterProvider
from opentelemetry.sdk.metrics.export import PeriodicExportingMetricReader
from opentelemetry.sdk.resources import (
    Resource,
    SERVICE_NAME,
    SERVICE_VERSION,
    DEPLOYMENT_ENVIRONMENT,
)
from opentelemetry.exporter.otlp.proto.http.trace_exporter import OTLPSpanExporter
from opentelemetry.exporter.otlp.proto.http.metric_exporter import OTLPMetricExporter
from opentelemetry.instrumentation.celery import CeleryInstrumentor

try:
    from opentelemetry.instrumentation.psycopg2 import Psycopg2Instrumentor
except ImportError:
    Psycopg2Instrumentor = None

try:
    from opentelemetry.instrumentation.redis import RedisInstrumentor
except ImportError:
    RedisInstrumentor = None


def initialize_telemetry():
    """Initialize OpenTelemetry for the worker service"""
    service_name = os.getenv("OTEL_SERVICE_NAME", "worker")
    service_version = os.getenv("WORKER_VERSION", "1.0.0")
    environment = os.getenv("DD_ENV", "demo")
    customer_tier = os.getenv("CUSTOMER_TIER", "premium")
    region = os.getenv("REGION", "us-east-1")
    otlp_endpoint = os.getenv("OTEL_EXPORTER_OTLP_ENDPOINT", "http://otel-collector:4317")

    # Resource attributes (tags)
    resource = Resource.create(
        {
            SERVICE_NAME: service_name,
            SERVICE_VERSION: service_version,
            DEPLOYMENT_ENVIRONMENT: environment,
            "customer_tier": customer_tier,
            "region": region,
        }
    )

    # Tracer provider
    tracer_provider = TracerProvider(resource=resource)
    trace_exporter = OTLPSpanExporter(endpoint=f"{otlp_endpoint}/v1/traces")
    tracer_provider.add_span_processor(BatchSpanProcessor(trace_exporter))
    trace.set_tracer_provider(tracer_provider)

    # Meter provider
    meter_provider = MeterProvider(
        resource=resource,
        metric_readers=[
            PeriodicExportingMetricReader(
                OTLPMetricExporter(endpoint=f"{otlp_endpoint}/v1/metrics"),
                export_interval_millis=10000,  # Export every 10 seconds
            )
        ],
    )
    metrics.set_meter_provider(meter_provider)

    # Instrument libraries
    CeleryInstrumentor().instrument()
    if Psycopg2Instrumentor:
        Psycopg2Instrumentor().instrument(enable_commenter=True, commenter_options={})
    if RedisInstrumentor:
        RedisInstrumentor().instrument()

    print(f"OpenTelemetry initialized for {service_name} v{service_version}")


# Initialize when module is imported
initialize_telemetry()
